aptos-move/aptos-vm/src/aptos_vm.rs

```diff
@@ -334,15 +341,49 @@ impl AptosVM {
         // Clear side effects: create new session and clear refunds from fee statement.
         let mut session = self
             .vm_impl
-            .new_session(resolver, SessionId::epilogue_meta(txn_data));
-        let fee_statement = AptosVM::fee_statement_from_gas_meter(txn_data, gas_meter, 0);
+            .new_session(resolver, SessionId::run_on_abort(txn_data));
 
-        match TransactionStatus::from_vm_status(
-            error_code.clone(),
-            self.vm_impl
+        if matches!(transaction_status, TransactionStatus::Keep(_))
+            && txn_data.fee_payer().is_some()
+            && txn_data.sequence_number == 0
+            && self
+                .vm_impl
                 .get_features()
-                .is_enabled(FeatureFlag::CHARGE_INVARIANT_VIOLATION),
+                .is_enabled(FeatureFlag::SPONSORED_AUTOMATIC_ACCOUNT_CREATION)
+        {
+            if let Err(err) =
+                create_account_if_does_not_exist(&mut session, gas_meter, txn_data.sender())
+                    .map_err(expect_no_verification_errors)
+                    .or_else(|err| {
+                        expect_only_successful_execution(
+                            err,
+                            &format!("{:?}::{}", ACCOUNT_MODULE, CREATE_ACCOUNT_IF_DOES_NOT_EXIST),
+                            log_context,
+                        )
+                    })
+            {
+                return discard_error_vm_status(err);
+            }
+        }
+
+        // Need to zero this out, because that's the legacy behavior. Otherwise we charge for txn
+        // size where we used to not.
+        let mut null_txn_data = txn_data.clone();
```

Comment:
Why not pass a size to `charge_change_set_and_respawn_session`? seems better than cloning the whole struct